### PR TITLE
Incremental Test Part 2 - TestAssetProjects

### PR DESCRIPTION
**Pull request by piotrpMSFT**

### CODE REVIEW

**Comment by piotrpMSFT** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build please.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build please.

**Comment by piotrpMSFT** [comment]
removed the failing centos test machine.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test OpenSUSE13.2 x64 Debug Build please.

File: build/test/TestAssetProjects.targets [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      <TestAssetProjects>
+        <BuildInputIncludeFilter>%(RelativeDir)**/*.*</BuildInputIncludeFilter>
+        <BuildInputExcludeFilter>%(RelativeDir)bin/**/*.*;%(RelativeDir)obj/**/*.*</BuildInputExcludeFilter>
+        <OutputDir>$([System.IO.Directory]::GetParent(%(Identity)))</OutputDir>
+        <ProjectPath>%(Identity)</ProjectPath>
+      </TestAssetProjects>
+
+      <TestAssetProjects>
+        <OutputName>$([System.IO.Path]::GetFileName(%(OutputDir)))</OutputName>
+        <OutputName Condition=" '%(RelativeDir)' == 'TestAssets/TestProjects/AppWithOutputAssemblyName/' ">MyApp</OutputName>
```
Comment (eerhardt) [inline build/test/TestAssetProjects.targets]: These seem weird - is it because the project.json specifies this name?

File: build/test/TestAssetProjects.targets [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        <OutputName Condition=" '%(RelativeDir)' == 'TestAssets/TestProjects/TestAppWithContentPackage/' ">AppWithContentPackage</OutputName>
+      </TestAssetProjects>
+
+      <TestAssetProjects>
+        <BuildOutput>%(TestAssetProjects.RelativeDir)bin/$(Configuration)/netcoreapp1.0/%(TestAssetProjects.OutputName).dll</BuildOutput>
```
Comment (eerhardt) [inline build/test/TestAssetProjects.targets]: Not all TestAssetProjects are 'netcoreapp1.0' are they?  Aren't some also desktop TFM?

File: build/test/TestAssetProjects.targets (hunk 1 above)
Comment (piotrpMSFT) [inline build/test/TestAssetProjects.targets]: That's exactly right. This mechanism allows us to rename the expected output.

File: build/test/TestAssetProjects.targets [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    <ItemGroup>
+      <NoAutoBuildTestAssets Include="TestAssets$(PathSeparator)TestProjects$(PathSeparator)**$(PathSeparator).noautobuild" />
+      <NoAutoBuildTestAssetProjects Include="%(NoAutoBuildTestAssets.RelativeDir)project.json" />
+
+      <TestAssetProjects Include="TestAssets$(PathSeparator)TestProjects$(PathSeparator)**$(PathSeparator)project.json" 
+                         Exclude="@(NoAutoBuildTestAssetProjects)"/>
+      <TestAssetProjects>
+        <BuildInputIncludeFilter>%(RelativeDir)**/*.*</BuildInputIncludeFilter>
+        <BuildInputExcludeFilter>%(RelativeDir)bin/**/*.*;%(RelativeDir)obj/**/*.*</BuildInputExcludeFilter>
+        <OutputDir>$([System.IO.Directory]::GetParent(%(Identity)))</OutputDir>
```
Comment (eerhardt) [inline build/test/TestAssetProjects.targets]: I think "OutputDir" is not a great name for this variable, since it is capturing the project directory, right?

File: build/test/TestAssetProjects.targets (hunk 2 above)
Comment (piotrpMSFT) [inline build/test/TestAssetProjects.targets]: https://github.com/dotnet/cli/blob/rel/1.0.0/build_projects/dotnet-cli-build/TestTargets.cs#L237

We moved the Desktop projects into DesktopTestProjects some time back.

File: build/test/TestAssetProjects.targets (hunk 1 above)
Comment (piotrpMSFT) [inline build/test/TestAssetProjects.targets]: I could create a seperate itemgroup, OutputNameOverrides, and then merge them. I chose this approach because there are less MSBuild concepts in play. What are your thoughts on the alternative?

File: build/Microsoft.DotNet.Cli.Prepare.targets [hunk 4]
```diff
@@ -4,7 +4,10 @@
   <UsingTask TaskName="GenerateBuildVersionInfo" AssemblyFile="$(CLIBuildDll)" />
   <UsingTask TaskName="CheckPrereqs" AssemblyFile="$(CLIBuildDll)" />
   <UsingTask TaskName="SetEnvVar" AssemblyFile="$(CLIBuildDll)" />
+  <UsingTask TaskName="DotNetBuild" AssemblyFile="$(CLIBuildDll)" />
```
Comment (eerhardt) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: It feels like these belong in a more global place that Prepare.targets.  Prepare doesn't using any of these new tasks, so it seems weird for it to be declaring them.

Maybe adding a new "global.props", or similar name, would be a good place for these UsingTask elements that are used across all our .targets files.

File: build/test/TestAssetProjects.targets (hunk 1 above)
Comment (eerhardt) [inline build/test/TestAssetProjects.targets]: IMO - this is the best approach until we actually have MSBuild working to build our projects.  I was just making sure I knew why this was doing what it was doing.  It is sad that we have to duplicate this in both the PJ and here, but I see nothing better in the short-term.

File: build/test/TestAssetProjects.targets (hunk 2 above)
Comment (eerhardt) [inline build/test/TestAssetProjects.targets]: But some target both right?

Or are we saying it is "good enough" to just check 'netcoreapp1.0' in those situations?

**Comment by eerhardt** [comment]
:shipit: after addressing the above comments.

File: build/test/TestAssetProjects.targets (hunk 3 above)
Comment (piotrpMSFT) [inline build/test/TestAssetProjects.targets]: Fixed.

File: build/Microsoft.DotNet.Cli.Prepare.targets (hunk 4 above)
Comment (piotrpMSFT) [inline build/Microsoft.DotNet.Cli.Prepare.targets]: Let's do this in a separate PR. For now, I've put them in the same place as other tasks coming from CLIBuildDll. Let's rationalize our Init/build.proj/Prepare story in one fell swoop.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test Windows_NT x64 Release Build  please.
